jxl/src/api/inner/codestream_parser.rs

```diff
@@ -186,28 +216,62 @@ impl CodestreamParser {
                 }
                 if self.sections.is_empty() {
                     // Go back to parsing a new frame header, if any.
-                    self.frame = None;
-                    return Ok(());
+                    self.process_without_output = false;
+                    if regular_frame {
+                        return Ok(());
+                    }
+                    continue;
                 }
             } else {
                 // Trying to read a frame or a file header.
                 assert!(self.frame.is_none());
                 assert!(self.has_more_frames);
+
+                match self.process_non_section(decode_options) {
+                    Ok(()) => {}
+                    Err(Error::OutOfBounds(_)) => {
+                        // If we need more data, that's expected - we'll get it on the next iteration
+                    }
+                    Err(e) => return Err(e),
+                }
+
+                if self.frame.is_some() {
+                    if self.has_visible_frame() {
+                        return Ok(());
+                    } else {
+                        self.process_without_output = true;
+                        continue;
+                    }
+                }
+
                 let available_codestream = box_parser.get_more_codestream(input)? as usize;
                 let c = self.non_section_buf.refill(
                     |buf| {
                         if !box_parser.box_buffer.is_empty() {
                             let c = box_parser.box_buffer.take(buf);
-                            Ok(box_parser.box_buffer.consume(c))
+                            Ok(c)
                         } else {
                             input.read(buf)
                         }
                     },
                     Some(available_codestream),
                 )?;
+
                 box_parser.consume_codestream(c as u64);
-                // TODO(veluca): skip returning if the current frame is not meant to be displayed.
-                return self.process_non_section(decode_options);
+
+                self.process_non_section(decode_options)?;
+
+                if self.decoder_state.is_some() && self.frame_header.is_none() {
```

Comment:
A comment on what is happening here would be nice.

Reply:
Done